lib/spack/spack/solver/concretize.lp

```diff
@@ -868,32 +858,44 @@ error(2, "'{0} target={1}' is not compatible with this machine", Package, Target
 %-----------------------------------------------------------------------------
 % Compiler semantics
 %-----------------------------------------------------------------------------
-compiler(Compiler) :- compiler_version(Compiler, _).
-
-% There must be only one compiler set per built node. The compiler
-% is chosen among available versions.
-{ attr("node_compiler_version", Package, Compiler, Version) : compiler_version(Compiler, Version) } :-
+% There must be only one compiler set per built node.
+{ node_compiler(Package, CompilerID) : compiler_id(CompilerID) } :-
     attr("node", Package),
     build(Package).
 
+% Infer the compiler that matches a reused node
+node_compiler(Package, CompilerID)
+  :- attr("node_compiler_version", Package, CompilerName, CompilerVersion),
+     attr("node", Package),
+     compiler_name(CompilerID, CompilerName),
+     compiler_version(CompilerID, CompilerVersion),
+     concrete(Package).
+
+% Expand the internal attribute into "attr("node_compiler_version")
+attr("node_compiler_version", Package, CompilerName, CompilerVersion)
+  :- node_compiler(Package, CompilerID),
+     compiler_name(CompilerID, CompilerName),
+     compiler_version(CompilerID, CompilerVersion),
```

Comment:
I can have a look at this in a following PR. It might help to do like you say and have a similar lookup on `error(2, ...)` involving compilers. But I can't tell for sure without trying it first (as anything clingo :sweat: )